### ISSUE TITLE
[RESEARCH] Figure out how best to manage Composer lockfile hygiene.

### ISSUE THREAD

**Issue by ndouglas** [department-of-veterans-affairs/va.gov-cms#10341]
## Description
The `composer.lock` file is not always up-to-date on the `main` branch.  I don't know why this is.  I think it may involve Dependabot updates, possibly merging Dependabot PRs into other PRs, etc.

I think I should go back through the commit history and run `composer install` at each commit and find what sort of characteristics are common among the commits where the `composer.lock` file is outdated.

## Acceptance Criteria
- [x] Check a bunch of commits to see what behaviors/circumstances cause the `composer.lock` file to get messed up.
- [x] Open up a followup issue to address (whether by documenting best practices, working on a CI fix, etc) (see #10502 ).

**Comment by ndouglas** [comment]
| Commit | Message | Clean/Unclean |
| -- | -- | -- |
| f6818bf174dbb | Upgrade to Drupal 9.4.5 (#10294) | Unclean |
| 35de6c8862 | Bump drupal/paragraphs from 1.14.0 to 1.15.0 (#10475) | Clean |
| bce2ca5466 | Bump va-gov/content-build from 0.0.3084 to 0.0.3085 (#10474) | Clean |
| 887b19ee | VACMS-10460: remove unused Lovell Tricare menu (#10463) | Unclean |
| e4d497d1 | toggle command outputs the wrong status (#10449) | Unclean |
| 275ab8dd | VACMS-10165: Add lovell documentation. (#10412) | Unclean |
| 8307bdb9 | Bump drupal/smart_date from 3.5.1 to 3.6.0 (#10362) | Unclean |
| fef4d12a | VACMS-10277: enhance facility services reports with additional fields (#10381) | Clean |
| 4de351b6 | Bump va-gov/content-build from 0.0.3082 to 0.0.3084 (#10424) | Clean |
| 1486fa3a | Bump drupal/embed from 1.5.0 to 1.6.0 (#10409) | Clean |
| 1ac98cce | VACMS-7600: Release content continuously during business hours (#10377) | Clean |
| 5b61e645 | VACMS-10374: Upgrade entityqueue to stable release. Add staged_content queue. (#10376) | Clean |
| 94e80382 | Bump drupal/memcache_admin from 2.4.0 to 2.5.0 (#10385) | Clean |
| 8ef542d0 | Committing ddev generated file. (#10404) | Clean |
| 053e652b | [docs] update Forms dataflow for times in ET | Clean |
| 20ec58fe | VACMS-10394 Adds config for Table Audit to be in the menu (#10395) | Clean |
| 0241af3e | Bump drupal/memcache from 2.4.0 to 2.5.0 (#10387) | Clean |
| aba25d89 | Revert "Adds services.local.yml for overriding service containers with local changes, and adds that to the .gitignore." | Clean |
| f4baa3cd | Adds services.local.yml for overriding service containers with local changes, and adds that to the .gitignore. | Clean |
| 06c704ef | Bump phpunit/phpunit from 8.5.28 to 8.5.29 (#10386) | Clean |
| 2a52ad53 | VACMS-10249: Directly require phpspec/prophecy. (#10393) | Clean |
| 278a5242 | Bump aws/aws-sdk-php from 3.233.0 to 3.234.0 (#10384) | Clean |
| 30110f72 | Bump va-gov/content-build from 0.0.3081 to 0.0.3082 (#10388) | Clean |
| ee9a29a0 | [docs] Forms issuetype: label update | Clean |
| 5a5fd8e3 | [docs] Created forms-bad-pdf-link-pw.md | Clean |
| 82bbbd85 | Update testing.md | Clean |
| 8b6024f1 | Update testing.md | Clean |
| 0a6c7865 | VACMS-10366: Removes no-reply@va.gov from Update Manager preferences. (#10367) | Clean |
| 9feb9b51 | Update issue templates | Clean |
| 98c34f49 | Bump drupal/coder from 8.3.15 to 8.3.16 (#10360) | Clean |
| 59ef8af3 | Bump va-gov/content-build from 0.0.3080 to 0.0.3081 (#10359) | Clean |
| 8e18300e | VACMS-10351 Reverts changes VACMS-9483 (#10353) | Clean |
| 24aaabf6 | [DOCS] Add contribution info to workflow. (#10352) | Clean |
| 0feac140 | VACMS-10147: Help text for Q&A Group Reusability Components (#10334) | Clean |
| 89915304 | Bump drupal/entity_block from 1.0.0 to 1.1.0 (#10330) | Clean |
| 24c0fed1 | Bump va-gov/content-build from 0.0.3078 to 0.0.3080 (#10331) | Clean |
| 9b532cc2 | Update system.site.yml (#10306)| Unclean |
| 781478e1 | VACMS-10184 Buttons reset selected fields on new content (#10250)| Unclean |
| f0c6f8c4 | VACMS 9996:  Reëvaluate packages with composer repository overrides (#10264)| Unclean |
| 0544a146 | VACMS-10249: Upgrade graphql explorer packages (undici and dset). (#10287) | Clean |
| 802ff7c4 | VACMS-7834: Defensive coding in theme_suggestions_paragraph_alter() hook implementations. (#10293) | Clean |
| c1b6834a | Update cms-qa-incidental.md | Clean |
| eda187a6 | Add jsonapi_menu_items and patch. (#10314) | Clean |
| fa0c4e84 | VACMS-10290: Upgrades Drupal to 9.3.21 [Security] (#10291)| Unclean |
| 2c9bbcb4 | Bump drupal/simplesamlphp_auth from 3.2.0 to 3.3.0 (#10304)| Unclean |
| 8edbad93 | Bump drupal/entity_reference_revisions from 1.9.0 to 1.10.0 (#10305) | Clean |
| ea0eaa7d | Bump va-gov/content-build from 0.0.3076 to 0.0.3078 (#10303) | Clean |
| f8e2c31f | Update cms-login.md| Unclean |
| 8ef83b38 | Update cms-login.md| Unclean |
| 5ac37476 | Update cms-login.md| Unclean |
| c6097dc6 | Update cms-login.md| Unclean |
| ad9387f8 | Bump datadog/dd-trace from 0.76.2 to 0.77.0 (#10037)| Unclean |
| 6d64a4cc | Bump knplabs/github-api from 3.7.0 to 3.8.0 (#10020)| Unclean |
| b8f87428 | Bump drupal/animated_gif from 2.0.1 to 2.0.2 (#10286) | Clean |
| 6bae3d85 | Bump va-gov/content-build from 0.0.3075 to 0.0.3076 (#10285) | Clean |
| 44b90dac | VACMS-9737: Implements Q&A reusability for Resource & Support (#9746) | Clean |
| 14a93e04 | Update README.md | Clean |
| df303a2f | Update qa.md | Clean |
| a8cd81b4 | Create qa.md | Clean |
| 6c08bde5 | Create cms-qa-incidental.md | Clean |
| 36273697 | Bump va-gov/content-build from 0.0.3074 to 0.0.3075 (#10244) | Clean |
| 010362bf | Bump drupal/raven from 4.0.0 to 4.0.1 (#10026) | Clean |
| aef6df18 | Bump geocoder-php/mapbox-provider from 1.3.0 to 1.4.0 (#10032) | Clean |
| 5164c623 | Bump aws/aws-sdk-php from 3.232.1 to 3.233.0 (#10204) | Clean |
| 07fec89d | VACMS-9956: add new section field to lovell menu items (#10110) | Clean |
| 2a4ef064 | Bump va-gov/content-build from 0.0.3071 to 0.0.3074 (#10203) | Clean |
| 0112fb89 | Bump drupal/address from 1.10.0 to 1.11.0 (#10128) | Clean |
| 573d2af6 | Add business hours calculation to all duration metrics (#10174) | Clean |
| 90003814 | Delete next-build component template | Clean |
| 84822b07 | update to remove NODE_EXTRA_CA_CERTS for ddev v1.20 (#10121) | Clean |
| f2ef3f0a | VACMS-9714: add scripts to update crisis number to 988 (#10044) | Clean |
| 78346486 | VACMS-10014: add promo_banner to the banners api response. (#10125) | Clean |
| 2427d78f | [docs] Update cms-team-and-sitewide-crew-member-onboarding.md | Clean |
| 3af6e1c7 | VACMS-10019: Improve content release resilience (#10095) | Clean |
| 2c826db1 | VACMS-10093 Updates instruction from 'Owner' to 'Section' (#10109) | Clean |
| b2e33832 | VACMS-9943: block lighthouse updates for Lovell Tricare section (#10082) | Clean |
| e02c33da | Update change-management.md | Clean |
| 7c3213f1 | Bump drupal/views_conditional from 1.4.0 to 1.5.0 (#10104) | Clean |
| 6481314b | Bump va-gov/content-build from 0.0.3070 to 0.0.3071 (#10103) | Clean |
| ec3666fe | Changes config to erase geocoding on failed look-up (#10101) | Clean |
| 604dd7ce | Bump gitonomy/gitlib from 1.3.5 to 1.3.6 (#10091) | Clean |
| b4235ecd | Bump va-gov/content-build from 0.0.3069 to 0.0.3070 (#10090) | Clean |
| 4ce44aaf | [docs] Update issue templates | Clean |
| b800d36d | Bump drupal/pfm from 1.2.0 to 2.0.0 (#10033) | Clean |
| 9e34f6e9 | Bump drupal/memcache_admin from 2.3.0 to 2.4.0 (#10076) | Clean |
| 3a0d10a8 | Bump drupal/memcache from 2.3.0 to 2.4.0 (#10077) | Clean |
| ed5f33de | Adds endpoint config patch from Geocoder module (#10056)| Unclean |
| 0ddc3a92 | Bump va-gov/content-build from 0.0.3066 to 0.0.3069 (#10074) | Clean |
| bcfdc3c9 | VACMS-10066 Swap floating patch for file patch. (#10069)| Unclean |
| 0a7d7f60 | [docs] Update issue templates | Clean |
| 1f350a0a | [docs] Update issue templates | Clean |
| 3d529dae | Bump aws/aws-sdk-php from 3.231.2 to 3.232.1 (#10047) | Clean |
| ba00ef33 | VACMS 10049: Updates text in Cypress test. (#10050) | Clean |
| dfa6d5ff | [docs] Update issue templates | Clean |
| c0852b71 | [docs] Creation of new cross team intake template | Clean |
| ec8755e3 | VACMS-0000: Correct promo banner allowed types to match expected singular values. (#10039) | Clean |
| d05937e2 | [docs] Update issue templates | Clean |
| e1343629 | [docs] Update issue templates | Clean |
| 66e45be8 | [docs] Update issue templates | Clean |
| 68ffa4ff | Adds config and tests for lovell-tricare (#10025) | Clean |

**Comment by ndouglas** [comment]
So we have a list of commits that seem to have introduced an "unclean" status:

- bcfdc3c9faf3089379d76dbe549c16944e334f12
- ed5f33dea6a664691c59ae1f2d23774be19adfc9
- 6d64a4cc4944318dd28fc40de4c7ba8b045973aa
- 2c9bbcb496d70308790b6f426f090016e0cfc5e7
- f0c6f8c4fca4e5354071be86715d5a5e079ccc3e
- 8307bdb9e81b40ec9cac0e131f70aa8fc0864649
- f6818bf174dbb6c34e2776a0f49545bb935ac81f

Curiously, these are created by human _and_ Dependabot contributors.  Why does Dependabot create broken commits? 🤔 

and a list of subsequent commits that **did not** reintroduce a "clean" status:

- f8e2c31
- 8ef83b3
- 5ac3747
- c6097dc
- ad9387f89e772660f87d74818d149fe09a035bf2
- 781478e124963d00cef02800f4e45dd1bd061c8e
- 9b532cc2641a7cb59a936ca83f7e1e4676f1c1c8
- 275ab8dd336a9b280b29fb2cf30b1e0ae7225cb1
- e4d497d13ee6a8e7a06e845b9bd395523adbdefe
- 887b19eed7c0a716347dd78401ef096fb5b7f6b8

Most of these are documentation commits or human contributions that didn't touch Composer.  But some of them are Dependabot contributions, like ad9387f89e772660f87d74818d149fe09a035bf2.  

Looking at #10037 closer, we can see that there're actually two commits involved: f41ead798eacd5adafb6814d6e9fa744a1ee329a, the commit approved, and ad9387f89e772660f87d74818d149fe09a035bf2, the commit actually merged.  The parent commit of f41ead798eacd5adafb6814d6e9fa744a1ee329a is b8f8742866948444801d3b082f24d9efafa8c1bd, while the parent commit of ad9387f89e772660f87d74818d149fe09a035bf2 is 6d64a4cc4944318dd28fc40de4c7ba8b045973aa.

6d64a4cc4944318dd28fc40de4c7ba8b045973aa is noted above as having introduced an "unclean" status; it did not update the `composer.lock` hash.  As a result, #10037's approved commit, f41ead798eacd5adafb6814d6e9fa744a1ee329a, did not show a merge conflict with `main`.  6d64a4cc4944318dd28fc40de4c7ba8b045973aa was merged into the #10037 branch cleanly, creating ad9387f89e772660f87d74818d149fe09a035bf2, which was then merged back into `main`.

I think this is the general pattern with these merges that don't repair the issue:
- Dependabot or a human changes the `composer.lock` file without updating the lock hash, creating an "unclean" state.
- One or more subsequent PRs that don't update the lock hash can be merged cleanly, but preserve the "unclean" state.
- At least one subsequent PR that updates the lock hash can be merged cleanly, and:
  - if it takes into account all previous changes in `main` at the time it is merged, then it will create a "clean" state
  - if it does not, then it will preserve the "unclean" state.

and a list of subsequent commits that **did** reintroduce a "clean" status:

- 0ddc3a92e171cf59b1380ef097bb690c871f4c03
- 3a0d10a854a10638ca9ca5cfd0a3f53c5befa39b
- ea0eaa7d21189780e8960661898393b3d77d4e5d
- eda187a62e349c5146cf23613713d18707132e4b
- 24c0fed100536e8d336e77ad577a2fcbabb3ce9f
- bce2ca5466d27442d378dfab9b46f83bc61dc87d

What do these have in common?

It appears that the approved commit and the merged commit have the same parent commits:
![Screen Shot 2022-08-29 at 10 37 29 AM](https://user-images.githubusercontent.com/1318579/187227063-6e341542-ce08-455b-af77-e1b01c00c7ea.png)

![Screen Shot 2022-08-29 at 10 37 36 AM](https://user-images.githubusercontent.com/1318579/187227022-5238b766-2d54-47ef-914c-4ed63be58b2a.png)

and thus the lock hash is guaranteed to be up-to-date.

So I think:

- We need to figure out why Dependabot is opening PRs that update packages but that don't update the lock hash.
- We need to require that PR branches be up-to-date, at least if they touch the Composer.lock file.
- We need to introduce an additional check on PRs to confirm that `composer install` indicates a clean lock file.

cc @timcosgrove @swirtSJW

**Comment by ndouglas** [comment]
After digging deeper, and actually reading some things, I discovered that I've been completely off-the-mark on how the `composer.lock` `content-hash` works.  I thought it was tracking installed packages, but actually does this:

https://github.com/composer/composer/blob/1.5.2/src/Composer/Package/Locker.php#L65-L102

So I have been misusing the tools in question.  Dependabot isn't creating broken PRs.  Some PRs involve changes to `composer.json`, and when they do, the `content-hash` is updated appropriately.  Sorry, based Dependabot, for having doubted you.

(Didn't Composer used to treat patches very differently? 🤔 )

To revise my above explanation of what is happening:

- Someone touches the `composer.json` file without updating the `composer.lock` file.  This is not Dependabot, but Dependabot's PRs are often related.  This is usually one of us removing or updating a patch to get the package installation to work cleanly.  This breaks the hash.  Since I can't read and misunderstood what was happening, I didn't run `composer update --lock`, which would've fixed the hash, prior to pushing that change to the PR branch.
- Any PR after that not repairing the hash will preserve the "unclean" state, including Dependabot PRs.
- At some point, the hash will get recalculated and create a "clean" state.

**TL;DR** I think all we need is a GHA that runs `composer validate`.  Very quick to do.
